### PR TITLE
#24 Codify your Developer VMs!

### PR DESCRIPTION
**Pull request by tknerr**
this is WIP, not ready for review yet... (remove links and rewrite last paragraphs)

### CODE REVIEW

**Comment by abeggchr** [comment]
@tknerr are you ready for a lead author review?

File: codify-your-developer-vms.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+the lifetime of the project.
+
+Finally, an automated developer VM should yield the following properties:
+
+* updatability - being able to update an existing VM with the push of a button from within the VM
+* testability - running automated tests in the VM to ensure that it works as expected
+* installability - the VM should distributed as a standard VirtualBox / VMware image
+* adaptability - developers should be able to easily adapt and improve the VM as the toolchain evolves
+* reproducibility - changes to the development environment are transparent, traceable and reproducible
+
```
Comment (abeggchr) [inline codify-your-developer-vms.md]: Gut und recht, aber der Versuch, die DevVM zu automatisieren ist in unserem Projekt grandios gescheitert. Es gab Probleme bei der MSSQL Installation (musste manuell ausgeführt werden weil nicht mit Chocolatey möglich), Office kann nicht installiert werden (weil die einen Teammembers eine Office 365 Lizenz haben, die anderen Office 2019 über MSDN), Windows an sich (schonmal eine Herausforderung mit der Lizenz) und dann und dann gibt es noch 1000 (oder sagen wir mal eine handvoll) Einstellungen welche doch noch manuell ausgeführt werden müssen: vs code plugins installieren, visual studio immer als admin ausführen, ...

Okay, genug Frust abgeladen. Den Artikel oben ist gut so und passt auch auf dem Artikel Level. Meine ganzen Umsetzungsfragen müssen ja da nicht beantwortet werden.

File: codify-your-developer-vms.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+## How can I automate it?
+
+The good news is that there are quite a few so-called "Configuration Management" tools out there, which essentially
+allow you to automate the provisioning of software and configuration on top of an existing operating system
+(in our case a fresh installation of the operating system of choice, within a VM image). 
+
+The bad news is that you have to choose one -- and in addition to the configuration management tools that do the
+provisioning, there is (gladly) also a multitude of tools for testing the configuration of such a (virtual) machine.
+
+Assuming you have a preference for either Ruby or Python, these are the tools you will likely end up with:
```
Comment (abeggchr) [inline codify-your-developer-vms.md]:
```suggestion
Depending on your preference, these are the tools you will likely end up with:
```

File: codify-your-developer-vms.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+* Pythonistas probably prefer
+  * Ansible - as the configuration management tool
+  * ansible-lint - for linting their Ansible roles
+  * TestInfra - for testing the Ansible-provisioned systems
+
+## I’m sold, sounds like pretty cool stuff! What is the fastest way to get there?
+
+It may sound a bit of effort to achieve this if you start from scratch, so that’s why we have a RUA with a
+template / skeleton project along with a tutorial for you. No excuses for doing it manually anymore! ;-)
```
Comment (abeggchr) [inline codify-your-developer-vms.md]:
```suggestion
template / skeleton project along with a tutorial for you.
```

File: codify-your-developer-vms.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+Not only for the consistency aspect, but also for reducing ramp-up times or being able to easily archive the
+development environment when the project enters the retirement phase.
+
+On regulated projects this is even more relevant, since you need to validate and verify the development toolchain
+and need to ensure that everyone uses exactly that defined toolchain.
+
+## OK, but why should I automate it?
+
+While we started out to craft these VMs manually (and with love) in the early days, we can nowadays easily
+automate the whole procedure with tools like Vagrant, Chef, Ansible and co. Applying automation and
```
[the remark this answers is not in the document]
Comment (damphyr) [reply codify-your-developer-vms.md]: "hardcore .Netties" should probably be "old MS gits nearing retirement" and they need to get up with the times then :) These all work with Windows and have done for a while :)

File: codify-your-developer-vms.md (hunk 1 above)
Comment (damphyr) [reply codify-your-developer-vms.md]: Und fuer jeden der aufgelistet Probleme eine Liste. OK, ich habe gesagt ich werde Torbens Artikel reviewen, nicht die Kommentare dazu...

**Comment by damphyr** [comment]
I am missing the "this is not all roses and lavender" part.
Two things to highlight: This is a compromise between consistency and performance. We err on the side of consistency because a fast flaky test is just going to flip-flop faster while we waste time looking for the differences.
And that the automation part is sometimes hard (especially on Windows) and is not just software installation but also configuration and settings like networking and firewalls and access rights etc.
These are not things everyone knows how to automate: so you have to know the tools *and* the platform real well. (this in response to @abeggchr 's mini Windows rant and experience, where I was going "yep, yep, solved that, had that, solved that one ...but it was tough" )

And then (and again this is how I think about this) I would like a juxtaposition between the manually crafted VM and the automated provisioning way: Theoretically we can enforce a manual provisioning of the VM *and* get the consistency benefits, but automation opens up other possibilities: scaling out is one - since we have coded our environment we can reproduce it and we can do it with little coupling to the underlying virtualization (meaning the VMs can be on desktop and on bare-metal hypervisors, or the provisioning can happen bare-metal as well, which then eliminates the performance penalty).
Which leads me to realize that we are here discussing two independent practices: Using a VM to encapsulate the development environment for the sake of consistency (something vital to our way of business but not universal) and using configuration management tools to put said environment configuration in code.
Combined they become a very powerful tool because they free you from the problems each one alone suffers from:
VM for encapsulation, script for reproducibility, documentation and scaling.

Also it might be worth clarifying that we have multiple choices: vagrant is our preferred tool for *developing* the VMs: quick iterations, carefree wipe and repeat, local. Nothing obliges the rest of the team to use vagrant when they can use the artifact created with vagrant - unless it fits.

And that we prefer the "vanila OS and then everything with scripts" approach to achieve  hypervisor independence. 
This is a bottomless well: I now have the urge to point out that we are also free to decide *how much* we automate (the concept of baseboxes for vagrant) and in how many stages (from .iso to basebox, to VM to multiple VM configurations for specialized roles). But this blows the article out of proportion.

File: codify-your-developer-vms.md (hunk 3 above)
Comment (damphyr) [reply codify-your-developer-vms.md]: Not everyone will be familiar with the RUA abbreviation

File: codify-your-developer-vms.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (tknerr) [reply codify-your-developer-vms.md]: Indeed that may sound like Linux only, but all of these tools do support Windows as well. By design they are actually OS agnostic. Will resolve that comment

File: codify-your-developer-vms.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+Some people hate it, others love it. However you put it, it is one sure-fire way to ensure that the
+whole development team uses a consistent environment and avoids the typical “works on my machine”
+issues we all know about. Especially for us, executing and delivering a multitude of projects (and engineers
+joining and leaving these projects more or less frequently), developer VMs have become an indispensable tool.
+Not only for the consistency aspect, but also for reducing ramp-up times or being able to easily archive the
+development environment when the project enters the retirement phase.
+
+On regulated projects this is even more relevant, since you need to validate and verify the development toolchain
+and need to ensure that everyone uses exactly that defined toolchain.
+
```
[the remark this answers is not in the document]
Comment (tknerr) [reply codify-your-developer-vms.md]: I understand that not everybody is a fan of running a VM in order to be able to work for a specific project, that's why I believe you can't or shouldn't force anyone to use it. In regulated projects, that might be easier as you might be forced by regulatory requirements to do that (i.e. the blame is not on you then ;)).

My observations so far:

* whenever you have Mac users, they are quite thankful to be provided with a VM if they work in a project where Windows is a requirement
* if you have people which frequently switch between projects, they are usually happy to have their toolchains / environments cleanly separated
* if you have people which are full time working for one project which has their preferred + installed operating system as a requirement, they rather work in their OS directly than in a VM
* disk space is an issue with developer VMs, that's definitely a cause for annoyance ;)

[the remark this answers is not in the document]
Comment (tknerr) [reply codify-your-developer-vms.md]: not sure though how I could write this in a more convincing style (and it won't help the haters anyway, probably ;)). 

@abeggchr if you have specific ideas on how to improve this I would happily take them though :)

File: codify-your-developer-vms.md (hunk 3 above)
Comment (tknerr) [reply codify-your-developer-vms.md]: @damphyr good point, I removed the reference to RUA

@abeggchr @damphyr I was a bit unsure about this last paragraph in general, considering that the article's content / text is there to stay, while the examples might be broken / non-functional at some point in time. Still they are useful. At least I reduced them to stuff that is hosted on our zuehlke Github account (which is fair to promote I guess).

Still I'm not 100% convinced about referencing these. What's your take on this?

**Comment by tknerr** [comment]
@damphyr good point about the "this is not all roses and lavender" part. Let me see how I can fit this in

File: codify-your-developer-vms.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (abeggchr) [reply codify-your-developer-vms.md]: If I had, I would have convinced more people to use a VM in our project.

**Comment by tknerr** [comment]
@damphyr unless you have further comments / feedback to the latest additions, I consider this ready for merge / lectorate

**Comment by tknerr** [comment]
Thanks @abeggchr ! 👍

**Pull request merged**